src/gt4py/next/program_processors/runners/dace_iterator/workflow.py

```diff
@@ -164,17 +199,41 @@ class Meta:
 
 
 def convert_args(
-    inp: stages.CompiledProgram,
+    inp: CompiledDaceProgram,
     device: core_defs.DeviceType = core_defs.DeviceType.CPU,
     use_field_canonical_representation: bool = False,
 ) -> stages.CompiledProgram:
-    sdfg_program = cast(CompiledSDFG, inp)
-    on_gpu = True if device == core_defs.DeviceType.CUDA else False
+    sdfg_program = inp.sdfg_program
     sdfg = sdfg_program.sdfg
+    on_gpu = True if device == core_defs.DeviceType.CUDA else False
 
     def decorated_program(
         *args, offset_provider: dict[str, common.Connectivity | common.Dimension]
     ):
+        if sdfg_program._lastargs:
+            # The scalar arguments should be replaced with the actual value; for field arguments,
+            # the data pointer should remain the same otherwise fast-call cannot be used and
+            # the args list needs to be reconstructed.
+            use_fast_call = True
+            scalar_args_table: dict[str, ctypes._SimpleCData] = {}
+            for pos, (param, arg) in enumerate(zip(sdfg.arg_names, args), start=1):
```

Comment:
I am very sure that this is not fully correct.
To use fast call you have to bring the arguments in the same order as the C interface, which is `<pointer-arguments>, <scalar.arguments>` each in alphabetical order. `sdfg.arg_names` is the names of the positional arguments, which can have a different order.

Furthermore, I do not see where you construct the arguments to the SDFG's init function.
As a side note here, calling the SDFG from Python uses a little known fact in the C standard:
```C
int foo();
{
        return 42;
}
```
In `C++` this is a function that does not take any arguments, however, in `C` the function takes an unknown number of arguments.
This works because all arguments are passed on the stack and it is the callers responsibility to put them there.
Thus you are not get an error.

The simplest way I know is to use `CompiledSDFG._construct_args()` which will do everything for you and update the internals.
The second possibility is to call `SDFG.arg_list()` which will give you the arguments in the corrected order.
Another option is the first time to simply use the normal `CompiledSDFG.__call__()` function, which will do it for you.

Reply:
It works because I am using `SDFG.signature_arglist`, which internally calls `SDFG.arg_list`, to retrieve the position of the arguments in the program ABI.

Reply:
One part might be incorrect, I see it now. The order of fields in `arg_names` is not necessarily the same as in the program ABI. I will fix this.
My answer applied to the order of scalars (represented as SDFG symbols): for scalar arguments I was already using the ABI order.